service/matching/matching_engine_test.go

```diff
@@ -2761,6 +2762,321 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
+	tqm, err := s.matchingEngine.getTaskQueuePartitionManager(context.Background(), ptq.Partition(), false)
+	s.NoError(err)
+	s.matchingEngine.unloadTaskQueuePartition(tqm)
+
+	// Loading the TQM for the same TQ partition
+	tqm_loaded, err := s.matchingEngine.getTaskQueuePartitionManager(context.Background(), ptq.Partition(), true)
+	s.NoError(err)
+	tqm_loaded_Impl, ok := tqm_loaded.(*taskQueuePartitionManagerImpl)
+	s.Equal(ok, true)
+
+	tlMgr, ok = tqm_loaded_Impl.defaultQueue.(*physicalTaskQueueManagerImpl)
+	s.Equal(ok, true)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), taskCount+1) // should read "taskCount + 1", not 1
+
+}
+
+// Concurrent tests for testing approximateBacklogCounter
+
+// TestConcurrentAddWorkflowTasksValidateBacklogCounter adds multiple workflow tasks concurrently to
+// persistence and checks if the backlog counter was updated accordingly.
+func (s *matchingEngineSuite) TestConcurrentAddWorkflowTasksValidateBacklogCounter() {
+	_, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const workerCount = 20
+	const taskCount = 100
+
+	var wg sync.WaitGroup
+	wg.Add(workerCount)
+
+	s.addWorkflowTasks(true, workerCount, taskCount, taskQueue, workflowExecution, &wg)
+	wg.Wait()
+
+	// validating the approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(taskCount*workerCount))
+}
+
+func (s *matchingEngineSuite) TestConcurrentAddCompleteWorkflowTasksValidateBacklogCounter() {
+	workflowType, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const workerCount = 20
+	const taskCount = 100
+
+	var wg sync.WaitGroup
+	wg.Add(2 * workerCount)
+
+	s.addWorkflowTasks(true, workerCount, taskCount, taskQueue, workflowExecution, &wg)
+	s.mockHistoryWhilePolling(workflowType)
+	s.pollWorkflowTasks(true, workerCount, taskCount, nil, taskQueue, &wg)
+
+	wg.Wait()
+	s.EqualValues(0, s.taskManager.getTaskCount(ptq))
+	totalTasks := taskCount * workerCount
+	persisted := s.taskManager.getCreateTaskCount(ptq)
+	s.True(persisted < totalTasks)
+
+	// validating the approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(0))
+}
+
+func (s *matchingEngineSuite) TestLesserNumberOfPollersThanTasks() {
+	workflowType, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const workerCount = 1
+	const taskCount = 5
+	const polledTasks = 4
+
+	var wg sync.WaitGroup
+	wg.Add(2 * workerCount)
+
+	s.addWorkflowTasks(true, workerCount, taskCount, taskQueue, workflowExecution, &wg)
+	s.mockHistoryWhilePolling(workflowType)
+	s.pollWorkflowTasks(true, workerCount, polledTasks, nil, taskQueue, &wg)
+
+	wg.Wait()
+	s.EqualValues(taskCount-polledTasks, s.taskManager.getTaskCount(ptq))
```

Comment:
is it right to expect the persisted count also be `taskCount-polledTasks`? does this check means we expect  `CompleteTasksLessThan` is called for all the polled tasks?

Reply:
Yes, eventually. As discussed aync, when we do have just 1 worker polling for these tasks, tasks should be taken out of order and eventually we should have 1 task in the backlog (in this case).

Does this not follow from our discussion yesterday?